### PR TITLE
First draft of PyBind11 bindings for ERS package

### PR DESCRIPTION
**Pull request by AlanTWatson**
First draft of bindings using pybind11. Hope there will be an update soon, will add once available.

There are a couple of scripts in the pytest directory that illustrate usage: 

pybind-test.py has simple examples of creating an issues directly using bound ers functions (prefixed ers: ) and using python-based helpers from erspy.py (analogues of the ers.py that came from ATLAS). 

erstest.py is the test script that came with the original ERS, modified to use these bindings. Everything runs, but some information isn't gathered properly yet.

### CODE REVIEW

File: pybindsrc/PyBind.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
+#include "ers/AnyIssue.hpp"
```
Comment (alessandrothea) [inline pybindsrc/PyBind.cpp]: I suggest to use `module.cpp` for the main file

File: pybindsrc/PyBind.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include "ers/ers.hpp"
+
+#include <exception>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+namespace py = pybind11;
+
+
+PYBIND11_MODULE(_daq_ers_py, liberspy) {
```
Comment (alessandrothea) [inline pybindsrc/PyBind.cpp]: Similarly, I would use `m` or `module` for the module object rather than `liberspy`

File: pybindsrc/erspy.cxx [hunk 3]
```diff
@@ -0,0 +1,226 @@
+#include <chrono>
```
Comment (alessandrothea) [inline pybindsrc/erspy.cxx]: This file is not picked up by the build system and it seems out of date. Please delete it.

File: pytest/pybind-test.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
+''' Script to test python bindings of ers methods
```
Comment (alessandrothea) [inline pytest/pybind-test.py]: This looks like a test script. If so it should be put in the `test/scripts/` folder

File: python/erspy.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
+'''Lots of imports. May not need them all, prune list later'''
```
Comment (alessandrothea) [inline python/erspy.py]: Is this file used?

Comment (AlanTWatson) [reply python/erspy.py]: It provides some of the python functionality provided by ers.py in the boost bindings, including a simple way for the user to just provide a message string and produce an ers Issue. Of course the pybind binding module allows you to produce Contexts and Issues directly, so it's not essential. We can develop this, or remove it if not needed, according to what fits our usage best.

File: pybindsrc/erspy.cxx (hunk 3 above)
Comment (AlanTWatson) [reply pybindsrc/erspy.cxx]: Ah, good catch - that's the old Boost binding!

Thanks for comments, I'll  work through them all.

File: pybindsrc/PyBind.cpp (hunk 1 above)
Comment (AlanTWatson) [reply pybindsrc/PyBind.cpp]: Updated in branch

File: pybindsrc/PyBind.cpp (hunk 2 above)
Comment (AlanTWatson) [reply pybindsrc/PyBind.cpp]: Updated in branch

File: pytest/pybind-test.py (hunk 4 above)
Comment (AlanTWatson) [reply pytest/pybind-test.py]: Updated in branch

File: pybindsrc/erspy.cxx (hunk 3 above)
Comment (AlanTWatson) [reply pybindsrc/erspy.cxx]: Updated in branch (also removed another copy I found!)

**Comment by AlanTWatson** [comment]
Applied all Alessandro's comments (so far!) in update to branch.

File: python/erspy.py (hunk 5 above)
Comment (AlanTWatson) [reply python/erspy.py]: Have an update now which includes most of the functionality this provided without requiring a separate include. Hence I've removed this file from the branch.

**Comment by plasorak** [comment]
Hi Alan, thanks for that. Am I correct in assuming that this snippet of code would throw an error?
```
import ers
class CantOpenFile( ers.PyIssue ):
    def __init__( self, fname, err, cause = None ):
        ers.PyIssue.__init__( self, 'Can not open "%s" file, error = %d' % ( fname, err ),
	    { 'filename' : fname, 'error' : err },
	    cause )
raise CantOpenFile( "test.py", 13, error )
```

The equivalent C++ code would be something like:
```
// in the header:
ERS_DECLARE_ISSUE(dfmodules,               ///< Namespace
                  DataStoreCreationFailed, ///< Type of the Issue
                  "Failed to create DataStore " << plugin_name << " with configuration "
                                                << conf,           ///< Log Message from the issue
                  ((std::string)plugin_name)((nlohmann::json)conf) ///< Message parameters
)
// in the code
throw DataStoreCreationFailed(ERS_HERE, type, conf, cexpt);
```

Could we have something like `ERS_HERE` in python, a function that does:
```
def ers_here():
    from inspect import currentframe, getframeinfo
    cf = currentframe()
    return cf.f_back.f_lineno, getframeinfo(cf.f_back).filename
```
So that we don't have to enter the filename and line number?

Also, I'm not able to run `pytest` in `test/scripts/`, how do you do it?

**Comment by AlanTWatson** [comment]
Hi Pierre,

The pybind-test.py script needs an argument. It's just some text for a message, so "python pybind11-test.py Hello" should work.

The code snippet is indeed intended just to throw an error (it's based on what was in the old boost-based release). I'm happy to change that (though stuck in meetings all afternoon, so may be this evening or tomorrow morning before I get round to it).

**Comment by plasorak** [comment]
Ah I was meaning [`pybind11_test.py`](https://github.com/DUNE-DAQ/ers/blob/atwatson/issue_19_feature_pybind_ers/test/scripts/pybind11_test.py) rather than `pybind-test.py`

**Comment by plasorak** [comment]
Ah nevermind, this is just a standard python script, sorry

**Comment by plasorak** [comment]
Ok, so to send message it is simply:
```
import ers
ers.pydebug(ers.message('a debug message'), 0)
ers.pylog(ers.message('a log message'))
ers.pyinfo(ers.message('an info message'))
ers.pywarning(ers.message('a warning message'))
ers.pyerror(ers.message('an error message'))
ers.pyfatal(ers.message('a fatal message'))
```
I think that's fine by me then. I initially got confused for the exceptions, but that looks ok too.

**Review by plasorak — APPROVED** [review]

**Comment by mroda88** [comment]
This should fix https://github.com/DUNE-DAQ/ers/issues/19

**Review by alessandrothea — APPROVED** [review]

**Pull request merged**